libraries/AP_BattMonitor/AP_BattMonitor_AD7091R5.cpp

```diff
@@ -0,0 +1,236 @@
+    
+
+    // calculate time since last current read
+    uint32_t tnow = AP_HAL::micros();
+    float dt = tnow - _state.last_time_micros;
+
+    // update total current drawn since startup
+    if (_state.last_time_micros != 0 && dt < 2000000.0f) {
+        // .0002778 is 1/3600 (conversion to hours)
+        float mah = _state.current_amps * dt * 0.0000002778f;
+        _state.consumed_mah += mah;
+        _state.consumed_wh  += 0.001f * mah * _state.voltage;
+    }
+
+    GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "Curr Pin : %d, raw_Value : %f final Value : %f _curr_amp_offset : %f  _curr_amp_per_volt : %f ",
+                                                curr_buff_pt, 
+                                                _data_to_volt(_analog_data[curr_buff_pt].data),
+                                                _state.current_amps,
+                                                _curr_amp_offset.get(),
+                                                _curr_amp_per_volt.get()
+                                                );         
+    // record time
+    _state.last_time_micros = tnow;
+}
+
+/**
+ * @brief Timmer function to read raw adc
+ * 
+ */
+void AP_BattMonitor_AD7091R5::_timer(){
+    _read_adc();
+}
+
+/**
+ * @brief read all four channels and store the results
+ * 
+ * @return true 
+ * @return false 
+ */
+bool AP_BattMonitor_AD7091R5::_read_adc() {
+    uint8_t data[AD7091R5_NO_OF_CHANNELS*2] = {0};
+    bool ret = _dev->transfer(nullptr, 0, data, sizeof(data));
+    if (ret) {
+        WITH_SEMAPHORE(sem);
+        for (int i=0; i<AD7091R5_NO_OF_CHANNELS; i++) {
+            uint8_t chan = AD7091R5_CH_ID(data[2*i]);
+            _analog_data[chan].data = ((uint16_t)(data[2*i]&AD7091R5_RES_MASK)<<8) | data[2*i+1];
+        }
+        // GCS_SEND_TEXT(MAV_SEVERITY_NOTICE, "%1.3f , %1.3f, %1.3f , %1.3f", _data_to_volt(_analog_data[0].data),_data_to_volt(_analog_data[1].data)
+        //                                                                 , _data_to_volt(_analog_data[2].data),_data_to_volt(_analog_data[3].data)); 
+    }
+
+    return ret;
+}
+
+/**
+ * @brief soft reset adc
+ * 
+ * @return true 
+ * @return false 
+ */
+bool AP_BattMonitor_AD7091R5::_reset(){
+    uint8_t reg = reg_conf_h | AD7091R5_RESET;
+    uint8_t data[3] = {AD7091R5_CONF_ADDR, reg, reg_conf_l};
+    return _dev->transfer(data, sizeof(data), nullptr, 0);
+}
+
+/**
+ * @brief power down adc
+ * 
+ * @return true 
+ * @return false 
+ */
+bool AP_BattMonitor_AD7091R5::_powerdown(){
+    uint8_t reg = (reg_conf_l & ~AD7091R5_CONF_PDOWN_MASK) | AD7091R5_CONF_PDOWN2;
+    uint8_t data[3] = {AD7091R5_CONF_ADDR, reg_conf_h, reg};
+    return _dev->transfer(data, sizeof(data), nullptr, 0);
+}
+
+/**
+ * @brief convert binary reading to volts
+ * 
+ * @param data 
+ * @return float 
+ */
+float AP_BattMonitor_AD7091R5::_data_to_volt(uint32_t data){
+    return (AD7091R5_REF*(float)data)/AD7091R5_RESOLUTION;
```

Comment:
```suggestion
    return (AD7091R5_REF/AD7091R5_RESOLUTION)*data);
```

It's not to important, but this is a faster version of the above. It's technically a bit less precise, but I would be surprised if you could show it causing problems on a 12 bit ADC.